pkg/utils/dataset/lifecycle/schedule.go

```diff
@@ -65,114 +79,144 @@ func AssignDatasetToNodes(runtimeInfo base.RuntimeInfoInterface,
-					log.Info("Node is skipped because it can't meet node selector terms", "node", node.Name)
-					continue
-				}
+		err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			nodeName := n.Name
+			node, err := kubeclient.GetNode(runtimeClient, nodeName)
+			if err != nil {
+				log.Error(err, "GetNode in AssignDatasetToNodes", "nodeName", nodeName)
+				return err
 			}
-		}
 
-		if !kubeclient.IsReady(node) {
-			log.Info("Node is skipped because it is not ready", "node", node.Name)
-			continue
-		}
-
-		if node.Spec.Unschedulable {
-			log.Info("Node is skipped because it is unschedulable", "node", node.Name)
-			continue
-		}
-
-		if len(node.Spec.Taints) > 0 {
-			tolerateEffect := toleratesTaints(node.Spec.Taints, dataset.Spec.Tolerations)
-			if tolerateEffect {
-				log.Info("The tainted node also can be scheduled because of toleration effects",
-					"node", node.Name,
-					"taints", node.Spec.Taints,
-					"tolerations", dataset.Spec.Tolerations)
-			} else {
-				log.Info("Skip the node because it's tainted", "node", node.Name)
-				continue
+			// No need to label the node if it fails the node filter
+			if !filterNode(*node, currentScheduledNodes, dataset, runtimeInfo, log) {
+				return nil
 			}
 
-		}
-
-		if !AlreadyAssigned(runtimeInfo, node) {
-			if !CanbeAssigned(runtimeInfo, node) {
-				log.Info("Node is skipped because it is not assigned and also can't be assigned", "node", node.Name)
-				continue
-			} else {
-				newScheduledNodes = append(newScheduledNodes, node)
-				log.Info("New Node to schedule",
-					"dataset", runtimeInfo.GetName(),
-					"node", node.Name)
+			if err = LabelCacheNode(*node, runtimeInfo, runtimeClient); err != nil {
+				return err
 			}
-		} else {
-			log.Info("Node is already scheduled for dataset",
+
+			// The candidate node is successfully labeled
+			log.Info("New Node to schedule",
 				"dataset", runtimeInfo.GetName(),
-				"node", node.Name)
-		}
+				"node", n.Name)
+			newScheduledNodes = append(newScheduledNodes, n)
+			currentScheduledNodes[n.Name] = n
+			return nil
+		})
 
-		currentScheduledNodes[node.Name] = node
+		if err != nil {
+			log.Error(err, "Scheduling node in AssignDatasetToNodes", "nodeName", n.Name)
+			return int32(len(currentScheduledNodes)), err
+		}
 	}
 
 	currentScheduleNum = int32(len(currentScheduledNodes))
 	newScheduleNum = int32(len(newScheduledNodes))
-	log.Info("Find node to schedule or scheduled for dataset",
+	log.Info("node scheduled for dataset",
 		"dataset", runtimeInfo.GetName(),
 		"currentScheduleNum", currentScheduleNum,
 		"newScheduleNum", newScheduleNum)
-	// 2.Add label to the selected node
 
-	for _, node := range newScheduledNodes {
-		err = LabelCacheNode(node, runtimeInfo, runtimeClient)
-		if err != nil {
-			return
+	return
+}
+
+// filterNode checks if the given runtime can be scheduled on the given node
+func filterNode(node corev1.Node,
```

Comment:
there is a file with name `node.go`
`filterNode` function associated with the file ?

Reply:
I think this func can move to node.go.

Reply:
This func is reverted.